src/translate/pddl_parser/parsing_functions.py

```diff
@@ -1,31 +1,142 @@
+SYNTAX_LITERAL = "(PREDICATE ARGUMENTS*)"
+SYNTAX_LITERAL_NEGATED = "(not (PREDICATE ARGUMENTS*))"
+SYNTAX_LITERAL_POSSIBLY_NEGATED = f"{SYNTAX_LITERAL} or {SYNTAX_LITERAL_NEGATED}"
+
+SYNTAX_PREDICATE = "(PREDICATE_NAME [VARIABLE [- TYPE]?]*)"
+SYNTAX_PREDICATES = f"(:predicates {SYNTAX_PREDICATE}*)"
+SYNTAX_FUNCTION = "(FUNCTION_NAME [VARIABLE [- TYPE]?]*)"
+SYNTAX_ACTION = "(:action NAME [:parameters PARAMETERS]? " \
+                "[:precondition PRECONDITION]? :effect EFFECT)"
+SYNTAX_AXIOM = "(:derived PREDICATE CONDITION)"
+SYNTAX_GOAL = "(:goal GOAL)"
+
+SYNTAX_CONDITION_AND = "(and CONDITION*)"
+SYNTAX_CONDITION_OR = "(or CONDITION*)"
+SYNTAX_CONDITION_IMPLY = "(imply CONDITION CONDITION)"
+SYNTAX_CONDITION_NOT = "(not CONDITION)"
+SYNTAX_CONDITION_FORALL_EXISTS = "({forall, exists} VARIABLES CONDITION)"
+
+SYNTAX_EFFECT_FORALL = "(forall VARIABLES EFFECT)"
+SYNTAX_EFFECT_WHEN = "(when CONDITION EFFECT)"
+SYNTAX_EFFECT_INCREASE = "(increase (total-cost) ASSIGNMENT)"
+
+SYNTAX_EXPRESSION = "POSITIVE_NUMBER or (FUNCTION VARIABLES*)"
+SYNTAX_ASSIGNMENT = "({=,increase} EXPRESSION EXPRESSION)"
+
+SYNTAX_DOMAIN_DOMAIN_NAME = "(domain NAME)"
+SYNTAX_TASK_PROBLEM_NAME = "(problem NAME)"
+SYNTAX_TASK_DOMAIN_NAME = "(:domain NAME)"
+SYNTAX_METRIC = "(:metric minimize (total-cost))"
+
+
+CONDITION_TAG_TO_SYNTAX = {
+    "and": SYNTAX_CONDITION_AND,
+    "or": SYNTAX_CONDITION_OR,
+    "imply": SYNTAX_CONDITION_IMPLY,
+    "not": SYNTAX_CONDITION_NOT,
+    "forall": SYNTAX_CONDITION_FORALL_EXISTS,
+}
+
+
+class Context:
+    def __init__(self):
+        self._traceback = []
+
+    def __str__(self) -> str:
+        return "\n\t->".join(self._traceback)
+
+    def error(self, message, item=None, syntax=None):
+        error_msg = f"{self}\n{message}"
+        if syntax:
+            error_msg += f"\nSyntax: {syntax}"
+        if item:
+            error_msg += f"\nGot: {item}"
+        raise ParseError(error_msg)
+
+    def expected_word_error(self, name, *args, **kwargs):
+        self.error(f"{name} is expected to be a word.", *args, **kwargs)
+
+    def expected_list_error(self, name, *args, **kwargs):
+        self.error(f"{name} is expected to be a block.", *args, **kwargs)
+
+    def expected_named_block_error(self, alist, expected, *args, **kwargs):
+        self.error(f"Expected a non-empty block starting with any of the "
+                   f"following words: {', '.join(expected)}",
+                   item=alist, *args, **kwargs)
+
+    @contextlib.contextmanager
+    def layer(self, message: str):
+        self._traceback.append(message)
+        yield
+        assert self._traceback.pop() == message
+
+
+def check_named_block(alist, names):
+    return isinstance(alist, list) and alist and alist[0] in names
+
+def assert_named_block(context, alist, names):
+    if not check_named_block(alist, names):
+        context.expected_named_block_error(alist, names)
+
+def construct_typed_object(context, name, _type):
```

Comment:
I agree with your point that `_type` should also be tested. After thinking about it a little bit, I don't see an issue with just doing the checks outside the function. Sure, one could argue that it makes sense to test this whenever trying to construct an object. However, within the constructing function, it would make more sense to assert this in my opinion.